resolver.go

```diff
@@ -14,34 +15,35 @@ import (
 	"github.com/aws/aws-sdk-go/service/servicediscovery"
 )
 
-type resolver struct {
-	mu       *sync.RWMutex
-	isClosed bool
+type serviceDiscovery interface {
+	DiscoverInstances(input *servicediscovery.DiscoverInstancesInput) (*servicediscovery.DiscoverInstancesOutput, error)
+}
 
+type resolver struct {
 	logger grpclog.LoggerV2
+	cc     grpcresolver.ClientConn
 
-	cc grpcresolver.ClientConn
-
-	ticker *time.Ticker
-
-	sd                 *servicediscovery.ServiceDiscovery
+	sd                 serviceDiscovery
 	namespace          string
 	service            string
 	healthStatusFilter string
 	maxResults         int64
+
+	ctx        context.Context
+	cancel     context.CancelFunc
+	ticker     *time.Ticker
+	resolveCmd chan struct{}
+	wg         sync.WaitGroup
 }
 
 func (c *resolver) ResolveNow(grpcresolver.ResolveNowOptions) {
-	locked := c.mu.TryLock()
-	if !locked { // already resolving
-		return
-	}
-	defer c.mu.Unlock()
-
-	if c.isClosed {
-		return
+	select {
+	case c.resolveCmd <- struct{}{}:
+	default:
 	}
+}
 
+func (c *resolver) cloudmapLookup() (*grpcresolver.State, error) {
```

Comment:
What about using verb tense for its name, like below?
Just `lookupCloudmap` would be great as well.

```suggestion
func (c *resolver) discoverInstances() (*grpcresolver.State, error) {
```